Semperis DirectoryServicesProtector connector for Azure Sentinel

ericlanteigne-semperis opened the pull request:
## Proposed Changes

  - Adding connector for Semperis DirectoryServicesProtector

ericlanteigne-semperis commented:
@v-jayakal @Amitbergman @dicolanl @ehudk-msft @ianhelle @juliango2100 @KobyKoren @laithhisham @liatlishams @liemilyg @lior-tamir @mgladi @nazang @NoamLandress @orco365 @oshezaf @oshvartz @petebryan @preetikr @sagamzu @sarah-yo @shainw @timbMSFT @Yaniv-Shasha @YaronFruchtmann 

Is anyone available to do a pull request for this? Please let me know if you are encountering any difficulties

v-jayakal commented:
@ericlanteigne-semperis , support for Events type data connector is in consideration for implementation. Will give you update as soon as that is finalized. Thanks.

v-admahe commented:
@ericlanteigne-semperis Could you please provide the sample data in the PR.

v-admahe commented:
@ericlanteigne-semperis Workbook is not working,it giving error in the queries.

![image](https://user-images.githubusercontent.com/73637076/124750107-b1dfc580-df42-11eb-9bfb-adc339a58f01.png)

Thanks,
Aditi

v-admahe commented:
@shainw could you please review the parser and analytics rule.

Thanks,
Aditi

ericlanteigne-semperis commented:
@v-admahe Could you please provide some details about what location and what format you would like the sample data to be? It causes validation errors based on where I put it so I want to make sure its done as you expect.

I believe the query for the dropdown options is considered to fail when there are no results returned. In this case since you don't have any events that contain this information it is not finding any values to add to the dropdown. Once you have the sample data they queries for these dropdowns will populate the values as expected. 

If you have some advice on how to make the dropdown simply show no values instead of showing a query failure that would be great. I will review documentation to see if I can find anything about that.

v-maudan commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+    "instructionSteps": [
+        {
+            "title": "", 
+            "description": ">This data connector depends on a parser based on a Kusto Function to work as expected. [Follow these steps](https://aka.ms/sentinel-SemperisDSP-parser) to create the Kusto Functions alias, **dsp_parser**",
```
Please update description as follows.
`"**NOTE:** This data connector depends on a parser based on a Kusto Function to work as expected [**dsp_parser**](https://aka.ms/sentinel-SemperisDSP-parser) which is deployed with the Azure Sentinel Solution."`

ericlanteigne-semperis replied:
Updated

v-maudan commented:
> @v-admahe Could you please provide some details about what location and what format you would like the sample data to have? It causes validation errors based on where I put it so I want to make sure its done as you expect.
> 
> I believe the query for the dropdown options is considered to fail when there are no results returned. In this case since you don't have any events that contain this information it is not finding any values to add to the dropdown. Once you have the sample data they queries for these dropdowns will populate the values as expected.
> 
> If you have some advice on how to make the dropdown simply show no values instead of showing a query failure that would be great. I will review documentation to see if I can find anything about that.

@ericlanteigne-semperis , For workbook, you can set default option for dropdown as 'All' and value as 'all' , so that even if there is no result queries will not fail. Can you please check attached updated workbook and let me know if it resolve this issue. Thanks! 
[SemperisDSPWorkbook.txt](https://github.com/Azure/Azure-Sentinel/files/6783554/SemperisDSPWorkbook.txt)

![image](https://user-images.githubusercontent.com/75478517/124903786-eebec180-e001-11eb-9e60-07ae9a3cddc3.png)

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+    "id": "SemperisDSP",
+    "title": "Semperis Directory Services Protector",
+    "publisher": "SEMPERIS",
+    "descriptionMarkdown": "Semperis Directory Services Protector data connector allows for the export of its Windows event logs (i.e. Indicators of Exposure and Indicators of Compromise) to Azure Sentinel in real time.\nIt provides a data parser to manipulate the Windows event logs more easily. The different workbooks ease your Active Directory seurity monitoring and provide different ways to vi ...
```
Good catch, thanks

ericlanteigne-semperis commented:
@v-maudan Thank you for the feedback. I've gone ahead and updated the text sections based on your recommendations.

I was testing the recommended change to the workbook to resolve the issue with dropdown queries failing when there is no data and I did encounter some issues. It seems that when using the new workbook the "All" selection doesn't work as expected. Once I had data available it would not display the data based on all values available, it would only work when selecting specific options. 

[Last 60 days with "All" as selections for subsequent dropdowns]
![AllShowingNoResults](https://user-images.githubusercontent.com/84532420/124934937-2febad00-dfd3-11eb-9b8f-b15eddc7954e.png)

[Last 60 days with specific framework selection and "All" for statuses]
![SelectionShowingNoResults](https://user-images.githubusercontent.com/84532420/124935034-4560d700-dfd3-11eb-923e-0a64873ec9fe.png)

Only once I select all of the options individually do I see the results the way I would when I used to have all selected
![SelectionWithAll3SelectedShowsData](https://user-images.githubusercontent.com/84532420/124935047-498cf480-dfd3-11eb-9127-e878a38792ad.png)

[Before recommended changes]
![PreviousSelectionShowingResults](https://user-images.githubusercontent.com/84532420/124935148-61fd0f00-dfd3-11eb-9bcf-99578dd749c4.png)

I can provide some sample data to help you reproduce the issue. Please let me know what format and where I can place that to avoid validations errors. 

_Note:
It's really odd but all of a sudden the workbook I submitted seems to be showing the "All" option when there is no data just like you are suggesting. I do recall seeing the query errors when there was no errors so it is throwing me off why I can't reproduce that anymore. I've tried clearing my cache incase something is being left behind after running the other workbook but it doesn't seems to change anything._

[Workbook with currently submitted code not showing query errors locally]
![FreshLoadDoesntShowQueryErrorsNow](https://user-images.githubusercontent.com/84532420/124935919-15fe9a00-dfd4-11eb-8c68-e4c8a6e623c9.png)

ericlanteigne-semperis commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+    "title": "Semperis Directory Services Protector",
+    "publisher": "SEMPERIS",
+    "descriptionMarkdown": "Semperis Directory Services Protector data connector allows for the export of its Windows event logs (i.e. Indicators of Exposure and Indicators of Compromise) to Azure Sentinel in real time.\nIt provides a data parser to manipulate the Windows event logs more easily. The different workbooks ease your Active Directory seurity monitoring and provide different ways to vi ...
+    "additionalRequirementBanner": "This data connector depends on a parser based on Kusto Function to work as expected. Follow the steps to use this Kusto Function alias **dsp_parser** in queries and workbooks. [Follow steps to get this Kusto Function>](https://aka.ms/sentinel-SemperisDSP-parser)  ",
```
Updated

ericlanteigne-semperis commented:
Unfortunately it looks like the DataConnectorValidations failed again with an issue I was encountering in the past. It seemed like the test had been updated since my validations had gone green but now they are failing again. However I wont be able to analyze the previous test results since they appear to have been cleaned up by the 10 day retention in project settings.

The issue is caused by the fact that we are using a new data type and the validations don't find the corresponding schema file. 

[DevOps pipeline showing validation error]
![image](https://user-images.githubusercontent.com/84532420/124937588-8659eb00-dfd5-11eb-9de4-2deb065fc009.png)

This is where the test substitutes the connectorCategoryinto the schema path and tries to find it. In our case the data type doesn't match those in the list and returns no connectorCategory. 
![image](https://user-images.githubusercontent.com/84532420/124937712-a5f11380-dfd5-11eb-9615-4149429c8958.png)

v-jayakal commented:
@ericlanteigne-semperis , can you attach few sample records in a json file in the PR? We will insert that data into a custom table and try using that for validation of the workbook and other queries using the custom table (instead of the Events table). Please refer to sample data files for CEF and SYslog here: https://github.com/Azure/Azure-Sentinel/tree/master/Sample%20Data. 
Let us know if you need more information. Thanks.

ericlanteigne-semperis commented:
@v-jayakal Okay I've added sample data into an Event directory under Sample Data since this is not Syslog or CEF related. Please let me know if you would like me to make an adjustment. 

@v-maudan I've included sample data that is used by the parser.

 I hope that helps.

v-maudan commented:
> @v-jayakal Okay I've added sample data into an Event directory under Sample Data since this is not Syslog or CEF related. Please let me know if you would like me to make an adjustment.
> 
> @v-maudan I've included sample data that is used by the parser.
> 
> I hope that helps.

@ericlanteigne-semperis , I have verified workbook with provided sample data , working fine. Thanks!

ericlanteigne-semperis commented:
@v-maudan Great to hear, thanks. 

I just need some advice about these failing validations at this point. 

DocumentsLinkValidation: I put the sample data in a new folder since its a new data type but the validations are failing saying "Documentation links should not include locale: Sample Data/Event/SemperisDirectoryServicesProtector.json,". 

DataConnectorValidations: Like I mentioned in my previous comment, this test is failing because the tests do not have logic for this data type and when it tries to find the corresponding schema, it fails to resolve. "Data Connector Validation Failed. File path: Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json. Error message: ENOENT: no such file or directory, open '.script/utils/schemas/_ConnectorSchema.json'"

ericlanteigne-semperis commented:
@v-maudan I am curious if you have some thoughts about this:
https://github.com/Azure/Azure-Sentinel/pull/2514#issuecomment-876472939

I was interested to make this improvement because I recall encountering the query errors for the dropdowns in the past when there was no results available but while testing your suggestions I was seeing some issues so I double checked what I had submitted and it was magically behaving the way we want it to. I don't recall it behaving this way so it seems like something is off but I wasn't able to make it go back to having an error when there was no data available. Are you able to reproduce the issue anymore?

v-rucdu commented:
> Unfortunately it looks like the DataConnectorValidations failed again with an issue I was encountering in the past. It seemed like the test had been updated since my validations had gone green but now they are failing again. However I wont be able to analyze the previous test results since they appear to have been cleaned up by the 10 day retention in project settings.
> 
> The issue is caused by the fact that we are using a new data type and the validations don't find the corresponding schema file.
> 
> [DevOps pipeline showing validation error]
> ![image](https://user-images.githubusercontent.com/84532420/124937588-8659eb00-dfd5-11eb-9de4-2deb065fc009.png)
> 
> This is where the test substitutes the connectorCategoryinto the schema path and tries to find it. In our case the data type doesn't match those in the list and returns no connectorCategory.
> ![image](https://user-images.githubusercontent.com/84532420/124937712-a5f11380-dfd5-11eb-9615-4149429c8958.png)

This has been fixed now.... Thanks!!!

v-rucdu commented:
> @v-maudan Great to hear, thanks.
> 
> I just need some advice about these failing validations at this point.
> 
> DocumentsLinkValidation: I put the sample data in a new folder since its a new data type but the validations are failing saying "Documentation links should not include locale: Sample Data/Event/SemperisDirectoryServicesProtector.json,".
> 
> DataConnectorValidations: Like I mentioned in my previous comment, this test is failing because the tests do not have logic for this data type and when it tries to find the corresponding schema, it fails to resolve. "Data Connector Validation Failed. File path: Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json. Error message: ENOENT: no such file or directory, open '.script/utils/schemas/_ConnectorSchema.json'"

@ericlanteigne-semperis : We fixed both the Validation errors and they are passing now.
DocumentsLinksValidation was failing because sample data had links to Azure Documentations with locale, have removed all such references and pushed latest changes.

DataConnectorValidation: We made a fix to allow the DataConnectorValidations for Events type.

Thanks!!!

v-rucdu commented on Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json:
```diff
@@ -0,0 +1,152 @@
+        },
+        "support": {
+            "name": "Semperis",
+            "link": "https:/https://www.semperis.com/contact-us/",
```
Please fix the typo in the url, **https:/** is repeated in the url.

Thanks!!!

ericlanteigne-semperis replied:
Great catch, thank you

ericlanteigne-semperis commented:
@v-maudan Please let me know if there is anything else missing. I believe I have addressed all of the issues we've discussed. 

If you have any ideas on that mystery I was mentioning earlier I am still interested to try to figure it out.

v-maudan commented:
> @v-maudan Please let me know if there is anything else missing. I believe I have addressed all of the issues we've discussed.
> 
> If you have any ideas on that mystery I was mentioning earlier I am still interested to try to figure it out.

@ericlanteigne-semperis , yes, All issues are resolved. :)

v-maudan commented:
> @v-maudan I am curious if you have some thoughts about this:
> [#2514 (comment)](https://github.com/Azure/Azure-Sentinel/pull/2514#issuecomment-876472939)
> 
> I was interested to make this improvement because I recall encountering the query errors for the dropdowns in the past when there was no results available but while testing your suggestions I was seeing some issues so I double checked what I had submitted and it was magically behaving the way we want it to. I don't recall it behaving this way so it seems like something is off but I wasn't able to make it go back to having an error when there was no data available. Are you able to reproduce the issue anymore?
> 
> [My version with 5 minute interval returning no results for the dropdowns]
> ![image](https://user-images.githubusercontent.com/84532420/125087177-490b6100-e09a-11eb-8d02-02ff23fec92f.png)

@ericlanteigne-semperis , If the query doesn't return any result then we are getting message to set paramter as mentioned in the screehot that is expected. Thanks!.

ericlanteigne-semperis commented:
Thank you @v-jayakal @v-maudan @v-rucdu @v-admahe for the help on this pull request!

And thank you @oshezaf and @v-admahe again for help on my previous one!
